packages/fast-tooling-react/src/form/utilities/navigation.ts

```diff
@@ -0,0 +1,381 @@
+        this.dataLocation = this.resolveNormalizeDataLocation(config.dataLocation);
+
+        this.setItemToNavigationMap(this.dataLocation);
+    }
+
+    /**
+     * Update the location, if not already part of the mapping
+     */
+    public updateDataLocation(
+        dataLocation: string,
+        callback: (updatedNav: NavigationItem[]) => void
+    ): void {
+        this.dataLocation = this.resolveNormalizeDataLocation(dataLocation);
+
+        if (
+            typeof get(this.navigationMap[dataLocation], "schemaLocation") === "undefined"
+        ) {
+            this.setItemToNavigationMap(this.dataLocation);
+        }
+
+        callback(this.get());
+    }
+
+    /**
+     * Update data at the current data location
+     */
+    public updateData(
+        data: unknown,
+        callback: (updatedNav: NavigationItem[]) => void
+    ): void {
+        this.config.data = cloneDeep(data);
+        this.updateChildrenLocations(); // children may have been added/removed
+
+        Object.keys(this.navigationMap).forEach((navigationMapKey: string) => {
+            if (typeof this.navigationMap[navigationMapKey].data !== "undefined") {
+                if (navigationMapKey === "") {
+                    this.navigationMap[""].data = this.config.data;
+                } else {
+                    this.navigationMap[navigationMapKey].data = get(
+                        this.config.data,
+                        this.navigationMap[navigationMapKey].dataLocation
+                    );
+                }
+            }
+        });
+
+        callback(this.get());
+    }
+
+    /**
+     * Gets an array of navigation items
+     */
+    public get(): NavigationItem[] {
+        const navigationItems: NavigationItem[] = this.resolveDataLocations(
+            this.dataLocation
+        ).map(
+            (dataLocation: string): NavigationItem => {
+                // return a previously resolved navigation item
+                if (typeof this.navigationMap[dataLocation] === "undefined") {
+                    this.setItemToNavigationMap(dataLocation);
+                }
+
+                return this.navigationMap[dataLocation];
+            }
+        );
+
+        return navigationItems.map(this.updateDefaultValues);
+    }
+
+    private updateChildrenLocations(): void {
+        this.reactChildrenDataLocations = getDataLocationsOfChildren(
+            this.config.schema,
+            this.config.data,
+            this.config.childOptions
+        );
+    }
+
+    private getDataLocationArray(dataLocation: string): string[] {
+        const dataLocations: Set<string> = this.getDataLocations(dataLocation);
+        const dataLocationIterator: IterableIterator<string> = dataLocations.values();
+        const dataLocationsArray: string[] = [];
+
+        // convert the set into of data locations into an array
+        for (let i: number = 0, size: number = dataLocations.size; i !== size; i++) {
```

Comment:
same here - you're iterating through twice - can you combine these?

Reply:
I'm using a Set here, so I need to convert it to an array